pennylane/qnodes/passthru.py

```diff
@@ -0,0 +1,145 @@
+
+    The advantages of this approach is that the qfunc can be differentiated using its AD framework
+    without requiring a separate method for computing the Jacobian, and that the internals
+    of the simulation are visible in the computational graph.
+
+    PassthruQNode requires some changes to the way other PennyLane components work:
+
+    1. :class:`Operator` must not do domain checking for its parameters, or it must let the ADT pass the check.
+    2. The simulator device must return the result as the ADT instead of plain Python/NumPy types.
+    3. Any output_conversion in :meth:`BaseQNode.evaluate` must be skipped.
+
+    Additionally, any array-like ADT needs to be able to handle (1) scalar multiplication,
+    (2) indexing/slicing, and possibly (3) iteration, as these are the things qfuncs expect of
+    array-like parameters.
+    """
+    def __init__(self, func, device, properties=None):
+        super().__init__(func, device, mutable=True, properties=properties)
+
+    def __repr__(self):
+        """String representation."""
+        detail = "<PassthruQNode: device='{}', func={}, wires={}>"
+        return detail.format(self.device.short_name, self.func.__name__, self.num_wires)
+
+    def _set_variables(self, args, kwargs):
+        # do nothing, since we do not use VariableRefs
+        pass
+
+    def _construct(self, args, kwargs):
+        """Construct the quantum circuit graph by calling the quantum function.
+
+        Like :class:`.BaseQNode._construct`, but does not use VariableRefs.
+
+        PassthruQNode does not have a Jacobian method, so it does not HAVE to use VariableRefs or scalar linear indexing of input parameters.
+        Two options:
+        1. Use VariableRefs anyway, re-use most BaseQNode methods.
+           Problem: after evaluating the VariableRefs, stacking sliced/indexed Tensors in Operation.parameters should somehow result in a Tensor, not an object array.
+        2. Do not use VariableRefs, call the qfunc each time :meth:`PassthruQNode.evaluate` is called (always mutable).
+           Problem: tensornet_tf requires variable_deps?
+
+
+        TODO rethink output_conversion? should require device to return things in a fixed form, but either as arrays or as AD Tensors, do conversion in interface (if necessary...)
+        """
+        # temporary queues for operations and observables
+        self.queue = []  #: list[Operation]: applied operations
+        self.obs_queue = []  #: list[Observable]: applied observables
+
+        # set up the context for Operator entry
+        if qml._current_context is None:
+            qml._current_context = self
+        else:
+            raise QuantumFunctionError(
+                "qml._current_context must not be modified outside this method."
+            )
+        try:
+            # turn off domain checking since PassthruQNode qfuncs can take any class as input
+            pennylane.operation.Operator.do_check_domain = False
```

Comment:
Not sure if this is the most elegant way to do things :thinking:
Can we not make the problematic code in `operation.Operator` more smart/flexible, rather than hand-coding this special case workaround?

Reply:
See above

Reply:
Something about the way github showed me updated comments meant I couldn't determine what "above" referred to here. Do you mean the earlier comment by @johannesjmeyer ?

Reply:
Yes

Hmm, looks like you can link to individual comments
https://github.com/XanaduAI/pennylane/pull/488/files#r375458913